### PR TITLE
[RFC] Keep expanded objects expanded

### LINKED ISSUE #74: Consistency in lists/objects
Thanks for your work, I'm VERY interested in something that deals with formatting in my place =)

One thing that struck me upon trying it out tonight, though, was that it does not handle "siblings" consistently. That is, given a list of, say, { id, name } objects; if one of them has a name large enough to reach the line limit, then this object, and this object alone, will be printed over 4 lines, while its siblings will be printed on 1 line. The "exploded" one will have a trailing comma, too, while obviously the one-liners will not.

Original `test.js`:
```
const myList = [
  {
    id: 0,
    name: 'Short',
  },
  {
    id: 1,
    name: 'Short Too',
  },
  {
    id: 3,
    name: 'Too Long For Your Line Width',
  },
];
```

Result (`$ prettier --bracket-spacing --trailing-comma --print-width 40 test.js`):
```
const myList = [
  { id: 0, name: "Short" },
  { id: 1, name: "Short Too" },
  {
    id: 3,
    name: "Too Long For Your Line Width",
  },
];
```

I think that in this case, the original is the better code: it's consistent, easily readable and parsable, at the only cost of taking a lot of vertical space (scrolling is cheap, though).

EDIT: removed JSX example and reworded the issue as per @vramana's comment.

### PR DESCRIPTION
**Pull request by vjeux**
Another attempt at solving the issue where objects are not expanded the way people expect. If there's any new line in the original source, it's going to expand it. This gives more control to the user in how the objects should be formatted.

Fixes #74

### CODE REVIEW

File: tests/flow/unicode/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -73,6 +73,8 @@ function utf16Length(str, pos) {
 exports[`test keys.js 1`] = `
 "({\'この事はつもり素晴らしいことさ\': \'35jL9V\'})
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-({ \"この事はつもり素晴らしいことさ\": \"35jL9V\" });
+({
+  \"この事はつもり素晴らしいことさ\": \"35jL9V\"
```
Comment (vjeux) [inline tests/flow/unicode/__snapshots__/jsfmt.spec.js.snap]: This is a bug in the flow parser, it incorrectly computes the length of the string with unicode characters inside.

**Comment by jlongster** [comment]
A lot of the tests changes are good example why I'm hesitant to do this. We should help the user simplify the code and a ton of those changes look like regressions to me. This seems like a place ripe for disagreements on a team and "nits" to start coming through to tell them to put an object on a single line because it's so small and fits inside the function call, etc.

This might be one of those things where I'd rather wait a while before doing, because we can't remove it after landing. I'd like to hear more from users about if this is an actual problem (only heard from a few that do, but many others may not want it)

**Comment by vjeux** [comment]
Yeah, I'm not really sure about it either. I'll keep sending various ideas to fix the problem :)

**Comment by jlongster** [comment]
Some of them seem like positive changes though:

screen shot 2017-01-27 at 1 58 38 pm

I suppose it does allow for consistency, so you might be onto something. Let's ask for more feedback about it.

**Comment by gaearon** [comment]
This looks much cleaner to me. When this kind of code evolves people end up adding more properties and you expand anyway at some point. Might as well do this right away. For types, you can always hoist them up for readability.

**Comment by rattrayalex** [comment]
Hmm this one feels a little concerning to me... Agreed it may be worth sitting on a bit...

**Comment by vjeux** [comment]
Some random thoughts on the subject:

The first attempts at solving the nits/formatting problem look like `eslint --fix`, where you would first look for bad patterns and then make them better automatically. The great thing is that if the user isn't breaking the rules, he can keep the formatting s/he wants. However, in practice this is very hard to implement properly outside of particular cases.

The second iteration is something like refmt where you reformat all the code with no exceptions. This is great and consistent but it is very inflexible. Developers want some control over how the code is formatted. #7 is an example that exploits the fact that empty lines are added before comments, so the person puts some random comments in order to make the code breathe.

I think that in order to succeed we need a middle ground. Where most of the code is automatically formatted but the user still has some control for cases where it cannot be always be automatically formatted in a reasonable way. We've done this successfully so far for empty lines, and objects may be a place where this would work as well.

**Comment by vjeux** [comment]
And, if we remove all the nits except for ones on objects, I still think the project would be super successful :)

**Comment by jlongster** [comment]
> And, if we remove all the nits except for ones on objects, I still think the project would be super successful :)

Alright. I don't have strong enough objections to hold this back. I'm fine moving forward with it. I don't want this to open the doors for other customizations based on input code; we should really try to restrict it. But I don't see this doing a lot of harm.

**Comment by vjeux** [comment]
Let's try this and revert it if it leads to people abusing it :)

**Pull request merged**